### PR TITLE
Refactor and add docs

### PR DESCRIPTION
**Pull request by charleskawczynski**
- Adds documentation
 - Adds structure to `src/`, fixes naming convention of files
 - Flattens and partition `test/`

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -1,6 +1,57 @@
...
+grav(::EarthParameterSet) = 5.0 # Defined for illustrative purposes
+@code_typed grav(param_set)
+```
+
+### Behavior 2) Run-time constants
+
+This behavior is used for parameters that **will** be tuned in the machine-learning layer. In this case, it makes sense to create a single binary image that can be reused for several different sets of CLIMAParameters as they are updated in the machine-learning layer. This is behavior is achieved by using composite types[^4] with fields to store/access data:
+
+```@example
+using CLIMAParameters:AbstractEarthParameterSet
```
Comment (vchuravy) [inline docs/src/index.md:35]:
```suggestion
using CLIMAParameters: AbstractEarthParameterSet
```

File: docs/src/index.md [hunk 2]
```diff
@@ -1,6 +1,57 @@
...
+ - The parameter is a function of only constants other CLIMAParameters and or constants
+
+In addition, CLIMAParameters have the flexibility of two important behaviors:
+
+### Behavior 1) Compile-time constants
+
+This behavior is used for parameters that **will not** be tuned in the machine-learning layer. Therefore, these parameters can be constant-propagated[^1] and constant-folded[^2] at compile time. This is behavior is achieved by leveraging Julia's type system, and only relying on singleton types[^3]:
+
+```@example
+using CLIMAParameters:AbstractEarthParameterSet
```
Comment (vchuravy) [inline docs/src/index.md:20]:
```suggestion
using CLIMAParameters: AbstractEarthParameterSet
```

File: docs/src/index.md [hunk 3]
```diff
@@ -1,6 +1,57 @@
...
+This behavior is used for parameters that **will** be tuned in the machine-learning layer. In this case, it makes sense to create a single binary image that can be reused for several different sets of CLIMAParameters as they are updated in the machine-learning layer. This is behavior is achieved by using composite types[^4] with fields to store/access data:
+
+```@example
+using CLIMAParameters:AbstractEarthParameterSet
+using InteractiveUtils
+import CLIMAParameters # imported for illustrative purposes
+using CLIMAParameters.Planet: grav
+
+struct EarthParameterSet{G} <: AbstractEarthParameterSet
+  _grav::G
```
Comment (vchuravy) [inline docs/src/index.md:41]:
```suggestion
  gravitiy::G
```

File: docs/src/index.md [hunk 4]
```diff
@@ -1,6 +1,57 @@
...
+end
+const param_set = EarthParameterSet(5.0)
+CLIMAParameters.Planet.grav(ps::EarthParameterSet) = ps._grav
```
Comment (vchuravy) [inline docs/src/index.md:44]:
```suggestion
CLIMAParameters.Planet.grav(ps::EarthParameterSet) = ps.gravitiy
```

**Comment by charleskawczynski** [comment]
@vchuravy the added docs are correct/precise?

**Comment by charleskawczynski** [comment]
@vchuravy bump!

**Pull request merged**